File: 2022/day_6/main.py
def main() -> None:
    with open("input.txt") as file:
        message: str = file.read()

    detected: bool = False
    packet: str = ""
    index: int = -1
    while not detected and index < len(message) - 3:
        index += 1
        packet: str = message[index:index + 4]
        if len(packet) == len(set(packet)):
            detected = True

    print(f"First packet with all unique characters: {packet} with index {index + 4}")

    detected: bool = False
    packet: str = ""
    index: int = -1
    while not detected and index < len(message) - 13:
        index += 1
        packet: str = message[index:index + 14]
        if len(packet) == len(set(packet)):
            detected = True

    print(f"First packet with all unique characters: {packet} with index {index + 14}")


if __name__ == '__main__':
    main()

File: 2022/day_6/main.go
package main

import (
    "io/ioutil"
    "fmt"
    "strings"
)

func hasUniqueChars(input string) bool {
    for _, c := range input {
        if strings.Count(input, string(c)) > 1 {
            return false
        }
    }
    return true
}

func getUniquePacketIndex(input string, length int) int {
    packet := input[:length]
    detected := false
    index := -1
    for !detected {
        index++
        packet = input[index:index+length]
        if hasUniqueChars(string(packet)) {
            detected = true
        }
    }
    return index + length
}

func main() {
    content, _ := ioutil.ReadFile("input.txt")

    fmt.Println(getUniquePacketIndex(string(content), 4))
    fmt.Println(getUniquePacketIndex(string(content), 14))
}
